Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/Search/ProductSearchService.cs

```diff
@@ -61,18 +61,21 @@ protected virtual IQueryable<ItemEntity> BuildQuery(ICatalogRepository repositor
             if (!criteria.CategoryIds.IsNullOrEmpty())
             {
                 var searchCategoryIds = criteria.CategoryIds;
-                searchCategoryIds = searchCategoryIds.Concat(repository.GetAllChildrenCategoriesIdsAsync(searchCategoryIds).GetAwaiter().GetResult()).ToArray();
-                //linked categories
-                var allLinkedCategories = repository.CategoryLinks.Where(x => searchCategoryIds.Contains(x.TargetCategoryId)).Select(x => x.SourceCategoryId).ToArray();
-                searchCategoryIds = searchCategoryIds.Concat(allLinkedCategories).Distinct().ToArray();
+
+                if (criteria.SearchInChildren)
+                {
+                    searchCategoryIds = searchCategoryIds.Concat(repository.GetAllChildrenCategoriesIdsAsync(searchCategoryIds).GetAwaiter().GetResult()).ToArray();
+                    //linked categories
+                    var allLinkedCategories = repository.CategoryLinks.Where(x => searchCategoryIds.Contains(x.TargetCategoryId)).Select(x => x.SourceCategoryId).ToArray();
```

Comment:
No need in linked categories, we just want to get products from child categories.